### ISSUE TITLE
ACP: add bool::select

### ISSUE THREAD

**Issue by orlp** [rust-lang/libs-team#468]
# Proposal

## Problem statement

It is a common theme in performance-oriented programming to want to select between two outcomes, but without changing which code runs as this invokes branch prediction. Branch prediction is great when it works but has two downsides:

1. It can be significantly slower than branchless code if the branch is unpredictable, like in sorting random data.
2. It can have inconsistent performance (even if faster on average), which can be a deal-breaker in latency-sensitive applications.

Currently in Rust if you want to have a good chance of something being branchless you have to write your code in a particular way, and even then it regularly fails to compile to branchless instructions, depending on the mood of the compiler.

## Motivating examples or use cases

There are tons of examples, I don't think I have to convince anyone familiar with performance-sensitive code of its use. Binary searches, sorting, filtering of data based on a condition, the applications are nearly endless. In performance-sensitive code, if you write an `if` where both sides of the `if` are cheap to evaluate and the condition is not 99%+ consistent there's a good chance a branchless `select` statement would be faster.

And of course there is the consistency argument even if the branch is 99%+ consistent in most cases, for example audio DSP code might want to ensure a kernel is always consistent in its speed to prevent audio crackling when a rare sequence of unpredictable samples come along, or high-frequency trading code may want to ensure it always has a response in a certain amount of nanoseconds.

## Solution sketch

I propose we add the following method to the `bool` primitive type:

```rust
/// Selects one of two values based on the value of self.
///
/// Unlike an `if` statement both arguments are always evaluated. The compiler
/// will try to emit machine code that avoids control flow for this selection,
/// taking the same amount of time regardless of whether self is true or false.
const fn select<T>(self, true_val: T, false_val: T) -> T;
```

This interface is similar to the [`std::simd::Mask::select`](https://doc.rust-lang.org/std/simd/struct.Mask.html#method.select) API, just for one value.

## Alternatives

There is the [`select_unpredictable`](https://doc.rust-lang.org/std/intrinsics/fn.select_unpredictable.html) intrinsic. However, it is an intrinsic so it should not really be used. There is no viable alternative, even writing things such as `[false_val, true_val][cond as usize]` can and do get rewritten by the compiler to branches, or just generate plain terrible code. One can reach for inline assembly but this pessimizes surrounding code and is not portable.

As for bikeshedding about the name, I considered the following alternatives, in order of preference:

- `select_predicated` (see [predication](https://en.wikipedia.org/wiki/Predication_(computer_architecture)))
- `select_eager`
- `select_branchless`
- `select_unpredictable` (as the intrinsic was named)

However, I prefer just the simple `select`. I think the fact it always eagerly evaluates both arguments (as functions do) is already self-documenting enough to readers to show it is not a control flow structure, but rather a selection operator between two pre-computed values. I think it is fairly natural that this then in turn compiles to a branchless select, even when simply reading `cond.select(a, b)`.

My reason for not favoring the current name of the intrinsic `select_unpredictable` is that (in my opinion) it confuses a problem with the solution. This can lead to confusing scenarios where the solution is applicable but the problem is not, e.g. someone could make an informed choice to use `select_unpredictable` while the boolean condition is in fact predictable - they just wanted consistency or avoid conditional jumps for other reasons.

## Links and related work

- [`select_unpredictable`](https://doc.rust-lang.org/std/intrinsics/fn.select_unpredictable.html)

**Comment by the8472** [comment]
This shouldn't go on the bool type itself. It's more something fit for the `std::hint` module

**Comment by orlp** [comment]
@the8472 It's not a hint, it's an actual operation.

**Comment by workingjubilee** [comment]
`std::hint` is not `std::intrinsics` by a different name, and this is not proposing to expose `select_unpredictable` directly.

**Comment by chorman0773** [comment]
I'd assume this would be allowed to generate a branch if no other codegen option exists.

**Comment by ChrisDenton** [comment]
It does seem a bit weird to have this as a method on bool. Maybe some code examples would change my mind but it feels too niche to not be a separate function or a wrapper type.

**Comment by orlp** [comment]
> I'd assume this would be allowed to generate a branch if no other codegen option exists.

Yes, the description should probably be expanded to explicitly mention this isn't a hard guarantee (even if compiling on a platform that has, e.g. `cmov`) and that it should not be relied upon for e.g. constant-time cryptography code.

**Comment by the8472** [comment]
Which then sounds like merely a hint to codegen...

**Comment by workingjubilee** [comment]
in the same sense that `unchecked_add` is, I suppose?

**Comment by the8472** [comment]
You have a point. But I'm not convinced yet. Unlike in SIMD llvm does have optimizations that turns cmovs into branches. So there's a difference between computing two values, picking one and then maybe the compiler transforms it depending on what it thinks is profitable and explicitly telling the compiler that you know better.
simd `Mask::select` doesn't have that because simd generally doesn't do branch prediction (maybe some stuff inside GPUs does?)

**Comment by chorman0773** [comment]
None of the functions in `core::hint` have any actual runtime behaviour (`unreachable_unchecked` and `assert_unchecked` are either UB or no-ops, `black_box` is an identity function, and `spin_loop` is a no-op). `bool::select` would have actual behaviour, so it definitely doesn't belong in `core::hint`.

**Comment by kennytm** [comment]
The proposed API does not fit into `std::hint`. However one could introduce it actually like a hint as

```rust
if std::hint::unpredictable(boolean) {
    true_value
} else {
    false_value
}
```

This should be "optimized" to `core::intrinsics::select_unpredicatable` if both `true_value` and `false_value` can be inlined and side-effect-free so that evaluating both branches are beneficial[^1]. This mirrors clang's [`__builtin_unpredictable` intrinsic](https://clang.llvm.org/docs/LanguageExtensions.html#builtin-unpredictable). Additionally, like `__builtin_unpredictable` it can be extended to support

```rust
match std::hint::unpredictable(integer) {
    0 => a,
    1 => b,
    3 => c,
    9 => d,
    27 => e,
    _ => etc,
}
```

OTOH in rust-lang/rust#120370 `core::intrinsics::{likely, unlikely}` are being reimplemented in terms of `cold_path()` because the intrinsics were broken, departing from hinting the condition and instead hinting the branches. 

According to the [LLVM docs][2] the `!unpredictable` metadata are supposed to "be attached to any branch or switch instruction", so it may be better aligned with LLVM to spell it as 

```rust
#[unpredictable]
if boolean { true_value } else { false_value }

#[unpredictable]
match integer { _ => etc }
```

[^1]: unfortunately LLVM will lose the `!unpredictable` metadata if we simply define `std::hint::unpredictable(boolean)` as `core::intrinsics::select_unpredicatable(boolean, true, false)`, but this may be just an LLVM bug.

[2]: https://llvm.org/docs/LangRef.html#unpredictable-metadata

**Comment by orlp** [comment]
@kennytm If `true_value` or `false_value` have side-effects such as panics, you'd need to write

```rust
let true_value = true_expr;
let false_value = false_expr;
#[unpredictable]
if boolean { true_value } else { false_value }
```
otherwise the Rust compiler is not allowed to write it using eager evaluation + a `cmov`/`csel`.

I'd really much rather just write

```rust
boolean.select(true_expr, false_expr)
```
which much closer matches my intent.

**Comment by workingjubilee** [comment]
@kennytm 

> According to the [LLVM docs](https://llvm.org/docs/LangRef.html#unpredictable-metadata) the !unpredictable metadata are supposed to "be attached to any branch or switch instruction", so it may be better aligned with LLVM to spell it as

That documentation is incorrect and outdated. It does in fact work on `select`. There is considerable code to propagate it and numerous tests to assure it does. It is possible the test coverage is not comprehensive enough, but that's compiler bugs for you.

- https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/CodeGenPrepare.cpp#L7395
- https://github.com/llvm/llvm-project/blob/main/llvm/lib/IR/IRBuilder.cpp#L1056
- https://github.com/llvm/llvm-project/blob/ac5a2010ad35a72de3e75a1883e2495345b92a73/llvm/test/CodeGen/X86/x86-cmov-converter.ll#L359-L430

**Comment by workingjubilee** [comment]
I meant the `select` LLVMIR instruction.

Did you verify `!unpredictable` disappears using Godbolt's LLVMIR view of Rust code?

I have found evidence to suggest it is **simply incorrect**.

**Comment by workingjubilee** [comment]
Oh. If you're just introducing the knowledge that trying to reiterate clang's implementation of `__builtin_unpredictable` (which resembles the `intrinsics::unlikely` API) is horribly broken, that is true. I am confused as to why you would mention the incorrect documentation, then.

**Comment by kennytm** [comment]
@workingjubilee if https://llvm.org/docs/LangRef.html is not correct then why it is not fixed 😅, the title literally said "LLVM 20.0.0git documentation" and reachable from https://llvm.org/docs/Reference.html#llvm-ir ← https://llvm.org/docs/#documentation ← https://llvm.org/. Also I don't see how that phrase I've quoted implies attaching `!unpredictable` to `select` is prohibited.



[godbolt] for reference.

[godbolt]: https://rust.godbolt.org/#z:OYLghAFBqd5TB8IAsQGMD2ATApgUWwEsAXTAJwBoiQIAzIgG1wDsBDAW1xAHIBGHpTqYWJAMro2zEHwAsQkSQCqAZ1wAFAB68ADIIBWMyozYtQ5AK4qSlNc3QkiI2q2wBhTIwscWIAMwATJSuADJELLgAct4ARrjkMgCslAAOmCqkTiweXj7%2BQWkZjiJhEdEccQl8yXa4DlliJGzkJDnevoG2uPbFLI3NJKVRsfFJtk0tbXmd1gND5ZVJAJS2mBbk6Nw8AKQBfmBg24kAQlKMmADuEOEk8eyMAPp0uGwk67gqS0cAIts6AIK7faHE7PV7vCBYci4B43cjhDLoT4/P6AgFAo7HFiYB4cUzAZgo9F7THhRjhXAQCIAN3iX0S3wA1KiAPQsxmRADyABV8CBGUQ6IyLrhGdgRGAeCRGWxsNhGSQUEQVMySScyRSqbhaeR6UyUPFRWRGXiANZGg3MgBsOmAmBw2xtrPZpnljp0MVl7sZdAsLHqIhVKDY1PCwDFgue0NEMpIJHhMQstxUlEZIRCADUALKSlW%2B/29E3xYBh52MiAUYXB246xkpNgqT7CpiMRlYCoUhUGjhij4pUgW0UHP0paHEBxsGLMAB0qJSFhiPpYjLtOAgMX5MXtjFTJH5Fk6Pv3gSWjIAtNs/PhGQeAsyAOzHVGMl%2BM6zYEAgOEIohIz%2B1BwHhHMdfyaKdKRiXdUzoL50XvX50WJPxMWxXF8UJBk5wXJdGU9bB103bdd2PIIjxvE9z0va9bwfJ8AVfRk2Q5Hk%2BS7IcwGA3BxzA5gBRVbFpVHTB62AV4uIFZdFVFBh7kZYTHA4IgAC9XiyZ9X0FN8SA/L9RHhFhERUf9ujqEggJYUcuNAydmHXXdLFwaCpDUU9tkfdSGJfEgPIfJlujUWifIYugfLchDAXgngVkYXhEkEXx%2BEETBeAAJSsaUVDWDZRSBARKD3Hg9CWFZTRARIdGMXhZEEDgZB0CqEr0Shkp4QQjIqgqisoOBYDQdt%2B2YKgaEhTAOAG0YWCIYAUBIRgAE8hCYW5yCMgjCsEGJwmaObeDyzb2HIObORifRTN2wR2y4URORYebdEEPBE2ANwziMxLKDwPEzGkdaPqIaF6lpN6mtwTQ6iTLY8pubp7uMIgYnIbaPDwWH4yIWr3p1Lc1G%2BXAvoJMN7pWOgTGAFQMyIXALk5FJWHOhRRAkKRuDkBnlDULRYcMPxjHxEBLGsOGYiMyAVnkrI3rPTlzzcXsnvCYRL2%2BHRzzPPHSEZRhGGpDgz3%2BpKdXhPARagdguBAKzekoakpAsLYAh0AIBGKroeiyFwWHcTx2iMUIKQWUY%2BFSdJMhEKZfCDwpQ5YeYRiqV3TIaCZWm9vIg4A3p%2BhaWOKkD8YBnDoxZmz/245kFZMvWTZy8qng4soRqkt4RkBZIdBGUm6bZrm8tCFIStcqWQROqilYDVlUYIBKsqKpinhqsoWq%2BHqhvYZatqQA6onot4AIarqhq194Yft8oUrl9n3g/H38rV9%2B9f8qJ7rEB6qA%2BpwAhiDIIbaAYZgzchoIYQjNJDSFZsA9mGhtC/UMEEEwZgQAPCQegBOAZfAQFcIXIOfsyhlx5lHXoWDg5FCyDnRYPMM5JwLqnCOqDM7JzIaMChyciHF0GKXXOCQ/AVyytXHYexwjoC8HgZkfg3Dvi3J4acKAqKoiBII4ROUxHvhuNI2RiF/gWBuIEB40o8Lrm3LhVMWjRA6OlDYG82iAi6J9K5dy9FXzQjeOQZcSDExLXCOZSy3EbIQVcn4AAYgqRk/IQp%2BDohFb4UVa710bs1XgstK7ZVVIEE%2B60XalVvnPBetVb5xIfu1R%2B6Tn7IHfmNJg8RqC0H6hUhI6AhH4j4PeHQQc/7LVWhuX6%2B1tr026YdY6p0HD00uqwEgN07q/UehYZ6r16afXQlsJqxAAaOCBrDUG4Nbj02hnPJq5IEZIw/qjeEGM8pY3SLgXG%2BNyQIOKSTNgZMKZUxpnTd6ECmZgPkBA1QUCuZGHgaAJBDwUH7JNmLFIvRJbSzVopEgZ4tY63PMAJqmBDZEGNvACuJk0Eey9rkWhODhicKMAQrIRDSUlA4YsdO2L6HUPxUXWlVCS64OJenFhNCi4MKpYHHhVduABB3nXeKR8eCMk0AADitGeK0sg2wAsZE06cOhpx8F7l/AeexBVpKKmPF4eAEhTzPjPWuC98nH1sJvIpura570Xia81rVrWj1rtfO%2BKKLUjxdjqDIzhZBAA

**Comment by Voultapher** [comment]
For what it's worth, this API gets a +1 from me. Efficient programs need ways to avoid expensive branch mispredicts. Generally branchless - or less confusingly named jumpless - programming is achieved via a small toolkit of building blocks:

- Conditional increment e.g. `num_x += cond as usize`, generates something like `inc or adc`.
- Perform both sides of the branch unconditionally and mask or ignore unwanted side.
- Value selection e.g. `let x = if cond { a } else { b }` generates something like `cmov or setb` with luck.
- Numeric logic, xor, bit and etc. only applicable for offsets or if the data handling isn't generic.
- SIMD masking.

Out of this non-exhaustive list of options one relies particularly heavy on compiler co-operation, branchless value selection.

**Comment by Voultapher** [comment]
@scottmcm I had a similar thought as well. There are different scenarios and and either `select` or `select_unpredictable` by themselves aren't necessarily a good fit for all of them. In theory we could add only `select_unpredictable`, or called `select` with the LLVM flags for unpredictable, and have `if cond { a } else { b}` favor generating branchy code, but I suspect that would break a non-trivial amount of existing optimizations, so probably best to go with new opt-in API and leave the current code-gen untouched.

When a programmer writes `if ...` it's not always clear what the intent is, and without PGO the compiler has to guess the mispredict penalty. LLVM tends to prefer branchless code while gcc is less aggressive about making such optimization. For example here https://cpp.godbolt.org/z/esxsdnvxf clang optimizes both functions to use cmov while gcc decides to use cmov in one case and a branch in the other. So giving developers a chance to express their intent in both cases seems like a good idea to me.

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/libs-api meeting.

We'd like to see this called `select_unpredictable`, with the justification that this would be a pessimization that you *shouldn't* use if your condition *is* predictable.

Otherwise: :+1: 🚢

**Comment by joboet** [comment]
I've implemented this in rust-lang/rust#133964.
